spec/lib/bookmark_manager_spec.rb

```diff
@@ -293,4 +293,36 @@ def notifications_for_user
       Notification.where(notification_type: Notification.types[:bookmark_reminder], user_id: bookmark.user.id)
     end
   end
+
+  describe ".toggle_pin" do
+    let!(:bookmark) { Fabricate(:bookmark, user: user) }
+
+    it "sets pinned to false if it is true" do
+      bookmark.update(pinned: true)
+      subject.toggle_pin(bookmark_id: bookmark.id)
+      expect(bookmark.reload.pinned).to eq(false)
+    end
+
+    it "sets pinned to true if it is false" do
+      bookmark.update(pinned: false)
+      subject.toggle_pin(bookmark_id: bookmark.id)
+      expect(bookmark.reload.pinned).to eq(true)
+    end
+
+    context "if the bookmark is belonging to some other user" do
+      let!(:bookmark) { Fabricate(:bookmark, user: Fabricate(:admin)) }
+      it "raises an invalid access error" do
+        expect { subject.toggle_pin(bookmark_id: bookmark.id) }.to raise_error(Discourse::InvalidAccess)
+      end
+    end
+
+    context "if the bookmark no longer exists" do
+      before do
+        bookmark.destroy!
+      end
+      it "raises an invalid access error" do
```

Comment:
that doesn’t look like an invalid access error, but a not found error

Reply:
Good point, bad copy/paste strikes again